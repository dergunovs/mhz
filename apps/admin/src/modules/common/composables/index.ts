import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
