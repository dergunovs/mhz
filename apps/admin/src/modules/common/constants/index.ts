export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
