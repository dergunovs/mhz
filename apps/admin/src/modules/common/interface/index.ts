import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
