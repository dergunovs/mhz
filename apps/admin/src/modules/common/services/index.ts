import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
