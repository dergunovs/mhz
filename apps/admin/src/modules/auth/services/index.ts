import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
