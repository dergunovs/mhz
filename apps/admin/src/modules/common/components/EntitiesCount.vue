<template>
  <div v-if="count" :class="$style.stats">
    <div :class="$style.top">
      <UiChart
        title="Base stats"
        :labels="count.base.labels"
        :datasets="count.base.datasets"
        data-test="entities-count-base"
      />

      <UiChart
        title="Products by categories"
        type="Pie"
        :labels="count.categories.labels"
        :datasets="count.categories.datasets"
        isShowLegend
        data-test="entities-count-product-categories"
      />
    </div>

    <UiChart
      title="Products by manufacturers"
      :labels="count.manufacturers.labels"
      :datasets="count.manufacturers.datasets"
      data-test="entities-count-product-manufacturers"
    />
  </div>
</template>

<script setup lang="ts">
import { UiChart } from 'mhz-ui';

import { getEntitiesCount } from '@/common/services';

const { data: count } = getEntitiesCount();
</script>

<style module lang="scss">
.stats {
  display: flex;
  flex-direction: column;
  gap: 64px;
}

.top {
  display: flex;
  gap: 64px;
}
</style>
