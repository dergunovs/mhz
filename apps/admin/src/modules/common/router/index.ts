import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
