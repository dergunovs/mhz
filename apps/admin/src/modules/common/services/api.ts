import axios from 'axios';

import { logout } from '@/auth/composables';

axios.defaults.baseURL = import.meta.env.VITE_API;

export const api = axios;

export function setAuthHeader(token: string) {
  api.defaults.headers.common['Authorization'] = `Bearer ${token}`;
}

export function deleteAuthHeader() {
  api.defaults.headers.common['Authorization'] = ``;
}

export function handleError(error: unknown) {
  if (axios.isAxiosError(error)) {
    if (error.response?.status === 403) logout();
    alert(error.response?.data.message);
  }
}
