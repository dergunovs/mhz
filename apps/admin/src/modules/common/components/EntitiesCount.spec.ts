import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IEntitiesCount } from 'mhz-contracts';
import { dataTest } from 'mhz-helpers';

import EntitiesCount from './EntitiesCount.vue';

import { wrapperFactory, mockQueryReply } from '@/common/test';
import { COUNT } from '@/common/fixtures';
import * as commonServices from '@/common/services';

const spyGetCount = vi.spyOn(commonServices, 'getEntitiesCount').mockReturnValue(mockQueryReply(COUNT));

const entitiesCountBase = dataTest('entities-count-base');
const entitiesCountProductCategories = dataTest('entities-count-product-categories');
const entitiesCountProductManufacturers = dataTest('entities-count-product-manufacturers');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(EntitiesCount, {});
});

enableAutoUnmount(afterEach);

describe('EntitiesCount', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(EntitiesCount)).toBeTruthy();
  });

  it('gets entities count', async () => {
    expect(spyGetCount).toBeCalledTimes(1);
  });

  it('passes props to child chart components', async () => {
    expect(wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountBase).vm.$props.labels).toEqual(
      COUNT.base.labels
    );

    expect(wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountBase).vm.$props.datasets).toEqual(
      COUNT.base.datasets
    );

    expect(
      wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountProductCategories).vm.$props.labels
    ).toEqual(COUNT.categories.labels);

    expect(
      wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountProductCategories).vm.$props.datasets
    ).toEqual(COUNT.categories.datasets);

    expect(
      wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountProductManufacturers).vm.$props.labels
    ).toEqual(COUNT.manufacturers.labels);

    expect(
      wrapper.findComponent<DefineComponent<IEntitiesCount>>(entitiesCountProductManufacturers).vm.$props.datasets
    ).toEqual(COUNT.manufacturers.datasets);
  });
});
