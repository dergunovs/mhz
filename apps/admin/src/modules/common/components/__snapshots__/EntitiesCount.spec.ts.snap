// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EntitiesCount > matches snapshot 1`] = `
"<div class="_stats_3f4722">
  <div class="_top_3f4722">
    <ui-chart-stub labels="Categories,Manufacturers,Managers,Customers,Orders" datasets="[object Object]" title="Base stats" type="Bar" isshowlegend="false" data-test="entities-count-base"></ui-chart-stub>
    <ui-chart-stub labels="Mousepad,Monitor,CPU,Keyboard,PSU,Cooler,Case,Mouse,Motherboard,GPU,SSD,RAM" datasets="[object Object]" title="Products by categories" type="Pie" isshowlegend="true" data-test="entities-count-product-categories"></ui-chart-stub>
  </div>
  <ui-chart-stub labels="MSI,ASRock,be quiet!,HyperX,Samsung,Fractal Design,Lian Li,Zalman,G.Skill,AMD,Thermaltake,Logitech,ASUS,GIGABYTE,Intel,Deepcool,Razer,Palit,Western Digital,Apacer" datasets="[object Object]" title="Products by manufacturers" type="Bar" isshowlegend="false" data-test="entities-count-product-manufacturers"></ui-chart-stub>
</div>"
`;
